freestyle/shared/src/main/scala/freestyle/free.scala

```diff
@@ -11,208 +11,163 @@ class free extends StaticAnnotation {
-        clsRestBody: List[Tree],
-        rootName: TypeName): List[(DefDef, ClassDef)] = {
-      for {
-        method <- clsRestBody filter {
-          case q"$mods def $name[..$tparams](...$paramss): FreeS[..$args]"     => true
-          case q"$mods def $name[..$tparams](...$paramss): FreeS.Par[..$args]" => true
-          case _                                                               => false
-        }
-        sc @ DefDef(_, _, _, _, tpe: AppliedTypeTree, _) = method
-        retType <- tpe.args.lastOption.toList
-        tparams = userTrait.tparams.tail ++ sc.tparams
-        args    = sc.vparamss.flatten //.filter(v => !v.mods.hasFlag(Flag.IMPLICIT))
-        //TODO user trait type params need to be added to the case class leave generated because its args may contain them
-        leaf: ClassDef = args match {
-          case Nil =>
-            q"""case class ${smartCtorNamedADT(sc.name.toTypeName)}[..${tparams}]()
-            extends $rootName[$retType]
-            """
-          case _ =>
-            q"""case class ${smartCtorNamedADT(sc.name.toTypeName)}[..${tparams}](..$args)
-            extends $rootName[$retType]"""
-        }
-      } yield (sc, leaf)
-    }
+    class Request(reqDef: DefDef) {
 
-    def mkSmartCtorsImpls(
-        typeArgs: List[TypeName],
-        adtRootName: TypeName,
-        scAdtPairs: List[(DefDef, ImplDef)]): List[DefDef] = {
-      for {
-        scAdtPair <- scAdtPairs
-        (sc, adtLeaf) = scAdtPair
-        cpType <- typeArgs.headOption.toList
-        injTpeArgs = adtRootName :: cpType :: Nil
-        ctor <- adtLeaf find {
-          case DefDef(_, TermName("<init>"), _, _, _, _) => true
-          case _                                         => false
-        }
-        args = sc.vparamss.flatten.collect {
-          case v /*if !v.mods.hasFlag(Flag.IMPLICIT)*/ => v.name
+      import reqDef.tparams
+
+      val reqImpl = TermName(reqDef.name.toTermName.encodedName.toString + "Impl")
+
+      // Name of the Request ADT Class
+      private[this] val Req: TypeName = TypeName(reqDef.name.toTypeName.encodedName.toString.capitalize + "OP")
+      private[this] val Res = reqDef.tpt.asInstanceOf[AppliedTypeTree].args.last
+
+      val params: List[ValDef] = reqDef.vparamss.flatten
+
+      def handlerCase: CaseDef  = {
+        val ReqC = Req.toTermName
+        if (params.isEmpty)
+          cq"$ReqC() => $reqImpl"
+        else {
+          // filter: !v.mods.hasFlag(Flag.IMPLICIT)
+          val ffs = params.map( v => q"l.${v.name}")
+          val uss = params.map( v => pq"_")
+          cq"l @ $ReqC(..$uss) => $reqImpl(..$ffs)"
         }
-        companionApply = adtLeaf match {
-          case c: ClassDef => q"${adtLeaf.name.toTermName}[..${c.tparams.map(_.name)}](..$args)"
-          case _           => q"new ${adtLeaf.name.toTypeName}"
+
+      }
+
+      def handlerDef: DefDef =
+        if (params.isEmpty)
+          q"protected[this] def $reqImpl[..$tparams]: $MM[$Res]"
```

Comment:
Given that the _effect_ trait (the one with the `@free` annotation) and the interpreter are part of a different class hierarchy, there should be no problems in their methods having the same names. And since the _Injection_ and the _Interpreter_ are to be defined in different contexts, there is no much room for confusion.